### PR TITLE
Feature/dc 693 improve api module

### PR DESCRIPTION
**Pull request by oleksii-symon-corva-ai**
[DC-693](https://corvaqa.atlassian.net/browse/DC-693)

### CODE REVIEW

**Review by ville — APPROVED** [review]

File: corva/network/api.py [hunk 1]
```diff
@@ -1,14 +1,31 @@
...
 from corva import settings
 
 
+@dataclass(frozen=True)
+class ApiResponse:
+    response: Response
+    data: Any = field(init=False)
+
+    def __post_init__(self):
+        object.__setattr__(self, 'data', self._load_data(response=self.response))
```
[the remark this answers is not in the document]
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py]: I've deleted this custom response class

File: corva/network/api.py [hunk 2]
```diff
@@ -1,14 +1,31 @@
...
-from requests import HTTPError, Session
+from requests import Response, Session
 from requests.adapters import HTTPAdapter
 from urllib3 import Retry
 
 from corva import settings
 
 
+@dataclass(frozen=True)
+class ApiResponse:
```
[the remark this answers is not in the document]
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py]: Deleted the `ApiResponse` class

File: corva/network/api.py [hunk 3]
```diff
@@ -1,60 +1,75 @@
...
+    def put(self, path: str, **kwargs):
         return self._request('PUT', path, **kwargs)
 
-    def delete(self, path, **kwargs):
+    def delete(self, path: str, **kwargs):
         return self._request('DELETE', path, **kwargs)
 
-    def _get_url(self, path):
+    def _get_url(self, path: str):
         if path.startswith(self.api_url) or path.startswith(self.data_api_url):
```
Comment (finomayato) [inline corva/network/api.py:73]: user already passed `api_url` and `data_api_url` into the constructor of this class. Why do we need to check if he is using the correct path? By possessing these values we have control over them and it's better to use them and not burden a user to provide a path in the format `f"{API_URL}/entity"`.

Though, I see a bigger problem here. It's unclear how the user will be using this class

**Review by finomayato — APPROVED** [review]
Code looks easier now!

I'm approving this PR, but my concern about the UI of this package still stands and I'd like to discuss it 

cc @jordanambra

File: corva/network/api.py [hunk 4]
```diff
@@ -68,16 +85,15 @@ def _get_url(self, path):
...
-         json=None,
-         params=None,
-         headers=None,
-         max_retries=None,
-         timeout=None,
-         asset_id=None,
-    ):
+         method: str,
+         path: str,
+         data: Optional[dict] = None,  # form-encoded data
```
[the remark this answers is not in the document]
Comment (jordanambra) [reply corva/network/api.py]: That should probably have a nicer method then--maybe an upload method.

[the remark this answers is not in the document]
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py]: Yeah, can we merge then?

[the remark this answers is not in the document]
Comment (jordanambra) [reply corva/network/api.py]: Yeah I think we can just make `data` the JSON param.

[the remark this answers is not in the document]
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py]: Do you mean this change `data: Optional[Json] = None` ? In this case user will need to json-encode his data for every request. In current case user passes a `dict`, which is converted to `json` automatically by `requests`.

File: corva/network/api.py (hunk 3 above)
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py:73]: Yeah, will delete this logic, it is obsolete.
User is supposed to use the class like this
```python
api = Api()
response = api.get(path='/api/v1/data/corva/dataset/')
```

Comment (finomayato) [reply corva/network/api.py]: We would like to take responsibility over switching api version too. So, if we are going update version - we need to roll-out new lib version

File: corva/network/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jordanambra) [reply corva/network/api.py]: That's all I mean. We don't need our old parameter that accepts POST key value pairs anymore. Just pass the dict straight to requests as the json parameter.

[the remark this answers is not in the document]
Comment (oleksii-symon-corva-ai) [reply corva/network/api.py]: Done

**Pull request merged**